pkg/kubelet/kubelet.go

```diff
@@ -1010,16 +1036,40 @@ func (kl *Kubelet) getClusterDNS(pod *api.Pod) ([]string, []string, error) {
 		return nil, nil, err
 	}
 
-	var dns, dnsSearch []string
+	var clusterDNS, clusterDomain string
+
+	// If cluster DNS support not specified on the command line, look it up
+	if kl.clusterDNS == nil || kl.clusterDomain == "" {
+		clusterDNS, clusterDomain, err = kl.lookupDnsService()
```

Comment:
Hmmm, ```Kubelet.makeEnvironmentVariables()``` (in the same file) does exactly the same thing, and shares exactly the same caller: ```Kubelet.GenerateRunContainerOptions()``` ... I fail to see where the additional locking is taking place. Could you point it out?

In addition, the Lister that both ```Kubelet.makeEnvironmentVariables()``` and ```Kubelet.lookupDNSService()``` are using caches the services. That's my limited understanding ... Does that make sense? So there should be no performance hit.

Lastly, it seems that lookup during kubelet startup would present a race condition. It would mean that Kubernetes nodes would need to be provisioned and started after the DNS service is available. This seems like an unnecessary limitation.